File: c/josephus-survivor.c
#include <stdio.h>

int josephus_survivor(int n, int k){

    int circle[n];
    int i;

    for(i=0; i<n;i++){
        circle[i] = i+1;
        // printf("%d", circle[i]);
    }

    int counter = 0;
    int t = 0;
    int j;
    int num_out = 0;

    while(num_out < n-1){
        // printf("hello");
        for(j=0; j<n; j++){
            if (circle[j] != 0){
                counter++;
            }
            if(num_out == n-1){
                break;
            }
            if (counter == k){
                circle[j] = 0;
                counter = 0;
                num_out++;
                // printf("%d", num_out);
            }
        }
    }

    int out = 0;
    for(i=0; i<n;i++){
        // printf("\n%d", circle[i]);
        if (circle[i] != 0){
            out = circle[i];
        }
    }


    return out;
}

int main(){
    int a = josephus_survivor(2545,1302);
    printf("%d", a);
    return 0;
}

File: python/5-kyu/josephus_survivor.py
import time
import numpy as np

start_time = time.time()


def josephus_survivor(n, k):
    # print(n,k)
    # return k%n

    circle = np.arange(1, n + 1)
    counter = 0
    t = 0

    while len(circle) > 1:
        # print(circle)
        circle_to_remove = np.zeros(len(circle))
        t = 0
        for i in circle:
            counter += 1
            if counter == k:
                circle_to_remove[t] = int(np.where(circle == int(i))[0])
                t += 1
                counter = 0
        circle_to_remove = circle_to_remove[0:t]
        for j in np.flipud(circle_to_remove):
            circle = np.delete(circle, int(j))

    print("--- %s seconds ---" % (time.time() - start_time))

    if np.any(circle):
        return int(circle[0])
    else:
        return n

    return 0
